core/system_impl.cpp

```diff
@@ -341,6 +347,11 @@ size_t SystemImpl::total_components() const
     return _components.size();
 }
 
+void SystemImpl::register_component_discovered_callback(discover_callback_t callback)
```

Comment:
What happens if the callback is registered after the component is discovered?

Reply:
In the current implementation the callback won't be called in that case.

Reply:
To be consistent with the system discovery, shouldn't we add that here, too? Or is it too complex because we then need to detect when a component times out?

Reply:
Only if the component timeout before the callback is registered. I'll take another look at it.

Reply:
In my understanding, [`auto res_pair = _components.insert(component_id);`](https://github.com/Dronecode/DronecodeSDK/pull/556/files#diff-780b2b438d1b4c2c3d32e3bc7b4f1d15R335) will be `true` only when a new component is inserted. And at that very moment, if a callback is registered (i.e. [`component_discovered_callback != nullptr`](https://github.com/Dronecode/DronecodeSDK/pull/556/files#diff-780b2b438d1b4c2c3d32e3bc7b4f1d15R337), then it will be called.

Which means that if events happen in this order:

1. User registers the callback
2. Component is discovered
3. Callback is called

Then everything is fine. But if the component has already been discovered when the user registers the callback, i.e.:

1. Component is discovered
2. Callback is `nullptr`, so it is not called
3. User registers the callback

Then the user just "missed" the callback, as the component will not be discovered again. However, I believe that in this scenario, the user actually wants to know that a component has been discovered, right?

Reply:
Yes, that could be true. In my case I know that I want to use the component discovery feature, so I place my callback registration appropriately. i.e. very early as in takeoff_land example.
I can see however that someone may use it differently.

Reply:
@JonasVautherin I've just push an update to fix this.

Reply:
Thanks, looking at it!